### PR TITLE
BUG 2276565: match csi-addons subscription channel to ocs-client-operator

### PR DESCRIPTION
**Pull request by rohan47**
this change will match csi-addons subscription channel with ocs-client-operator in 4.15

### CODE REVIEW

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:480]: instead of label can we go with known name which we use for subscription?

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:449]: dont we need to watch for subscription as well?

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:61]: who is adding the labels? can we go with names as we know it already.

**Review by Madhu-1 — COMMENTED** [review]
@rohan47 i think the final decision was this is the fix going to be present in all version we don't need to send this PR to 4.15 isn't it?

File: controllers/clusterversion_controller.go
Comment (rohan47) [reply controllers/clusterversion_controller.go:61]: The labels are getting added by olm as per my understanding
and regarding known name, the name for csi-addons is `odf-csi-addons-operator-stable-4.15-redhat-operators-openshift-marketplace` where the channel part gets included so its simple to use the label

File: controllers/clusterversion_controller.go
Comment (rohan47) [reply controllers/clusterversion_controller.go:449]: I am just fetching the subscription and changing the channel if it doesn't match ocs-client-operator as my intention here was to change the channel for csi-addons if it doesn't get upgraded

File: controllers/clusterversion_controller.go
Comment (iamniting) [reply controllers/clusterversion_controller.go:480]: The name can be different, But the package name is unique identity under the spec and we use that in the odf.

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/clusterversion_controller.go
Comment (rohan47) [reply controllers/clusterversion_controller.go:61]: making the change to package name instead of labels as Nitin suggested

**Comment by leelavg** [comment]
> @rohan47 i think the final decision was this is the fix going to be present in all version we don't need to send this PR to 4.15 isn't it?
- yes, we do need a change in 4.16, even though the final outcome will be same the codepath diverged a lot and decision was to have a separate fix for 4.15 alone and 4.16 will be a separate one. That's what I remember.

**Comment by leelavg** [comment]
@rohan47 thinking again, I agree w/ Nitin, we might want to go with checking the package name, able to use existing labels set by olm seems good but it has edge cases.

Btw, pls do update necessary rbac.

**Comment by rohan47** [comment]
Testing the changes

File: controllers/clusterversion_controller.go
Comment (iamniting) [reply controllers/clusterversion_controller.go:450]: We should watch for subscription if not already, otherwise how this code will be triggered automatically if there is a change in the subscription

Comment (rohan47) [reply controllers/clusterversion_controller.go:451]: added watch for subscriptions, verifying the code changes

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:111]: Add a predicate to trigger reconcile for only subscription in operator namespace

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:468]: what if we dont find these subscriptions? can we sync and have common idea for this PR and https://github.com/red-hat-storage/ocs-client-operator/pull/159 as the internal on how we are handling the subscription remains same?

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:477]: you need to do update here which is missing

Comment (rohan47) [reply controllers/clusterversion_controller.go:477]: Added

Comment (leelavg) [reply controllers/clusterversion_controller.go:477]: Comment should be slightly rephrased, the reasoning we want stable-4.15 is because of a direct merge to release-4.15 channel and this code shouldn't be back ported and from 4.16 anyways the logic is already changed.

Comment (rohan47) [reply controllers/clusterversion_controller.go:478]: Updated the comment to
```
//the channel of csiAddons subscription here is stable-4.15 because of a direct merge to release-4.15
//this code shouldn't be back ported and from release-4.16 the logic is changed
```

File: controllers/clusterversion_controller.go
Comment (leelavg) [inline controllers/clusterversion_controller.go:483]: what is this check for? seems redundant to me.

Comment (rohan47) [reply controllers/clusterversion_controller.go]: removed

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:477]:
```suggestion
// The channel of csiAddons subscription here is stable-4.15 because of a direct merge to release-4.15. This code shouldn't be backported, and from release-4.16, the logic has changed.
```

add a space between `//` and the word

**Review by Madhu-1 — COMMENTED** [review]
small nit, LGTM.

File: controllers/clusterversion_controller.go
Comment (rohan47) [reply controllers/clusterversion_controller.go]: done

**Comment by rohan47** [comment]
Getting this on local

```$ make lint
docker run --rm -v /home/mrudraia/redhat/ocs-client-operator:/app:Z -w /app golangci/golangci-lint:v1.54.2 golangci-lint run ./...
Unable to find image 'golangci/golangci-lint:v1.54.2' locally
v1.54.2: Pulling from golangci/golangci-lint
de4cac68b616: Pull complete 
d31b0195ec5f: Pull complete 
9b1fd34c30b7: Pull complete 
2f744fb50e06: Pull complete 
661ee9827c7b: Pull complete 
fa5c1764d694: Pull complete 
824755c07d59: Pull complete 
1edf93b66cce: Pull complete 
Digest: sha256:2082f5379c48c46e447bc1b890512f3aa9339db5eeed1a483a34aae9476ba6ee
Status: Downloaded newer image for golangci/golangci-lint:v1.54.2
level=warning msg="[runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused."
level=warning msg="[runner] The linter 'structcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused."
level=warning msg="[runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused."```

**Comment by Madhu-1** [comment]
>  level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
  level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
  level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."

@rohan47 can you update the ci-lint from v2 to v4 and see if that works?

**Comment by rohan47** [comment]
> > level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
> > level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
> > level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."
> 
> @rohan47 can you update the ci-lint from v2 to v4 and see if that works?

Where do I update ci-lint from v2 to v4?
Do I update the golint image tag in `hack/make-project-vars.mk:GO_LINT_IMG_TAG ?= v1.54.2`

**Comment by Madhu-1** [comment]
> > > level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
> > > level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
> > > level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."
> > 
> > 
> > @rohan47 can you update the ci-lint from v2 to v4 and see if that works?
> 
> Where do I update ci-lint from v2 to v4? Do I make the v2 to v4 change as per https://github.com/red-hat-storage/ocs-client-operator/blob/main/.github/workflows/golangci-lint.yaml Do I update the golint image tag in `hack/make-project-vars.mk:GO_LINT_IMG_TAG ?= v1.54.2`

add a commit to change v2 to v4 above, i don't think GO_LINT_IMG_TAG is used in CI to install specific version of golangci-lint

**Comment by leelavg** [comment]
Hmm, we need to do some more housekeeping then. CI seems to be using latest lint always irrespective of Makefile vars or release branch.

Maybe it's better to use a specific version for branch and latest in main branch 🤔

**Comment by rohan47** [comment]
> > > > level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
> > > > level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
> > > > level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."
> > > 
> > > 
> > > @rohan47 can you update the ci-lint from v2 to v4 and see if that works?
> > 
> > 
> > Where do I update ci-lint from v2 to v4? Do I make the v2 to v4 change as per https://github.com/red-hat-storage/ocs-client-operator/blob/main/.github/workflows/golangci-lint.yaml Do I update the golint image tag in `hack/make-project-vars.mk:GO_LINT_IMG_TAG ?= v1.54.2`
> 
> add a commit to change v2 to v4 above, i don't think GO_LINT_IMG_TAG is used in CI to install specific version of golangci-lint

Done, can you please approve the workflow

**Comment by leelavg** [comment]
Gave a try locally, first time it failed similar to CI but later it got succeeded. @rohan47 pls add `-v` as well to https://github.com/red-hat-storage/ocs-client-operator/blob/release-4.15/.github/workflows/golangci-lint.yaml#L42 and let's se if CI shows any extra failures.

**Comment by rohan47** [comment]
> Hmm, we need to do some more housekeeping then. CI seems to be using latest lint always irrespective of Makefile vars or release branch.
> 
> Maybe it's better to use a specific version for branch and latest in main branch 🤔

Added `-v` in the args and based on the following 
```Warning: Unexpected input(s) 'skip-go-installation', valid inputs are ['version', 'args', 'working-directory', 'github-token', 'only-new-issues', 'skip-cache', 'skip-pkg-cache', 'skip-build-cache', 'install-mode']```
removed `skip-go-installation`

File: .golangci.yaml
Comment (Madhu-1) [inline .golangci.yaml:10]: @rohan47 it looks like the changes in this file are enough to pass the CI, can you please revert the changes in the golangcilint.yml file in action?

Comment (rohan47) [reply .golangci.yaml:10]: done

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
> Hmm, we need to do some more housekeeping then. CI seems to be using latest lint always irrespective of Makefile vars or release branch.
> 
> Maybe it's better to use a specific version for branch and latest in main branch 🤔

Yes i think we should stick to versions instead of depending on the latest version which is not meant for release branches

**Comment by Madhu-1** [comment]
@leelavg if everything looks good, please place an `approve` comment from your side.

**Comment by leelavg** [comment]
/lgtm
/approve

Finally, I was waiting for CI.

**Review by leelavg — APPROVED** [review]

**Pull request merged**
